### ISSUE TITLE
Erroneous rounding upon assignment to array element

### ISSUE THREAD

**Issue by markrichardson** [numpy/numpy#7730]
Ubuntu 14.04
Python 2.7.11

```python
In [1]: import numpy as np

In [2]: np.__version__
Out[2]: '1.10.4'

In [3]: v = 4 - 0.1

In [4]: np.array([v])
Out[4]: array([ 3.9])

In [5]: a = np.array([1]); a[0] = v; a
Out[5]: array([3])
```

**Comment by seberg** [comment]
The behaviour is of course correct, `a` is autotyped to be an integer array, so the float is cast to an integer on assignment. This could have a warning in principle for the unsafe cast, but it is expected and has to work that way since numpy arrays are (normally) typed.

**Comment by charris** [comment]
Note that numpy doesn't round in these cases, it truncates. @seberg I thought this was deprecated, but it doesn't seem so.

**Comment by seberg** [comment]
There is only a deprecation for the in place ufuncs, maybe we should think about making this deprecate too?

**Comment by charris** [comment]
@seberg Sounds like a good idea, we should raise it on the list.

**Comment by njsmith** [comment]
https://mail.scipy.org/pipermail/numpy-discussion/2016-June/075679.html

**Comment by njsmith** [comment]
Sounds like everyone is in favor, so I renamed the bug report. I guess concretely the desired change should follow the same path we used for in-place operations? Which IIRC was that we switched it to same_kind casting, with a deprecation period.

**Comment by seberg** [comment]
Yeah, would have to be a pretty long deprecation just like there. I am a little worried that the workaround with copyto or casting the right hand side is sometimes a bit annoying. But I suppose that it should be rare that a bit of explicit casting/more convolved stuff does not hurt and the error can easily tell you the solution.

**Comment by sumeetsk** [comment]
What makes this especially unexpected / counter-intuitive is that python does not behave this way.

![image](https://user-images.githubusercontent.com/983247/49534420-1a480100-f887-11e8-81e9-b9d90b196659.png)

**Comment by mattip** [comment]
Overflow and truncation seem to be mentioned often as issues. Maybe we could have a `verbose` mode that would report overflow and truncation as warnings, at a cost in performance

**Comment by stevengj** [comment]
(Note that this issue and #8733 are basically duplicates.)
